vamb/reclustering.py

```diff
@@ -0,0 +1,319 @@
+"""
+The following code is a modification of a k-means based reclustering algorithm first published at https://github.com/BigDataBiology/SemiBin
+The original code is distributed under MIT License.
+"""
+
+import os
+import subprocess
+import tempfile
+import sys
+import contextlib
+from typing import IO
+from sklearn.cluster import KMeans
+import numpy as np
+from collections import defaultdict
+import pandas as pd
+
+
+def log(string: str, logfile: IO[str], indent: int = 0):
+    print(("\t" * indent) + string, file=logfile)
+    logfile.flush()
+
+
+def fasta_iter(fname, full_header=False):
+    '''Iterate over a (possibly gzipped) FASTA file
+
+    Parameters
+    ----------
+    fname : str
+        Filename.
+            If it ends with .gz, gzip format is assumed
+            If .bz2 then bzip2 format is assumed
+            if .xz, then lzma format is assumerd
+    full_header : boolean (optional)
+        If True, yields the full header. Otherwise (the default), only the
+        first word
+
+    Yields
+    ------
+    (h,seq): tuple of (str, str)
+    '''
+    header = None
+    chunks = []
+    if hasattr(fname, 'readline'):
+        op = lambda f,_ : f
+    elif fname.endswith('.gz'):
+        import gzip
+        op = gzip.open
+    elif fname.endswith('.bz2'):
+        import bz2
+        op = bz2.open
+    elif fname.endswith('.xz'):
+        import lzma
+        op = lzma.open
+    else:
+        op = open
+    with op(fname, 'rt') as f:
+        for line in f:
+            if line[0] == '>':
+                if header is not None:
+                    yield header,''.join(chunks)
+                line = line[1:].strip()
+                if not line:
+                    header = ''
+                elif full_header:
+                    header = line.strip()
+                else:
+                    header = line.split()[0]
+                chunks = []
+            else:
+                chunks.append(line.strip())
+        if header is not None:
+            yield header, ''.join(chunks)
+
+
+normalize_marker_trans__dict = {
+    'TIGR00388': 'TIGR00389',
+    'TIGR00471': 'TIGR00472',
+    'TIGR00408': 'TIGR00409',
+    'TIGR02386': 'TIGR02387',
+}
+
+
+def get_marker(hmmout, fasta_path=None, min_contig_len=None, multi_mode=False, orf_finder = None, contig_to_marker = False):
```

Comment:
This can return a list, a dict, or a defaultdict. That's a mess. Can it be cleaned up so its return type is predictable, or at least sensical?